### ISSUE TITLE
Generalise job_id retrieval by regex matching

### ISSUE THREAD

**Issue by guillaumeeb** [dask/dask-jobqueue#132]
See @willirath PR #131.

I believe this is a really good idea to use regex matching and make it configurable. We should extend this behaviour to other schedulers. Ideally, we would not need to implement `_job_id_from_submit_output` in JobQueueCluster implementations, and only have it in core.py, retrieving configuration value using scheduler name.

**Comment by lesteve** [comment]
In SGE qsub, there is a command-line argument to just output the job-id (so in SGE there is no need for a regex). Maybe this option exists in other schedulers too?

I am not too sure about making it configurable, I am hoping that we can cover all use cases with a per-scheduler regex that lives in our code rather than saying "if your cluster is a bit weird you can always tweak your dask config".

Ultimately it comes down to how much variation we expect from the submission command output depending on scheduler version or cluster setup. Maybe each cluster defines a qsub wrapper and customise its output? @guillaumeeb you probably have a way better feeling for this than I have so your input on this variability would be more than welcome!

**Comment by guillaumeeb** [comment]
Actually I did not expect variability for PBS, so this is a first example of bad first feeling.
However, with the Torque fork, older PBS versions, there may be some other unexpected outputs.

But my first concern was for LSF, which has a very verbose output that seem to vary according to configuration. And for LSF, I'm almost sure there is no way of outputting only JOB_ID.

Maybe Slurm has some different outputs too, don't know though.

One of the major point of using configurable regex, is to reduce python code by removing all `_job_id_from_submit_output` methods. So the per-scheduler regex lives in a config, and there is one single line of code.

But maybe this is overkill...

**Comment by lesteve** [comment]
I had a quick look at the infamous ipyparallel [file](https://github.com/ipython/ipyparallel/blob/master/ipyparallel/apps/launcher.py) mentioned in #49. You can look for `job_id_regexp` to figure out which regexp is used by controllers to parse the job-id from the submission command output. The controllers use `r'\d+'` for finding the job id, except HTCondor which uses `r'(\d+)\.$'`).

Maybe we can just follow what ipyparallel has done and hope that it covers most of schedulers variations out there.

I am still somewhat unsure about making it configurable ... I would still be in favour of not making it configurable, waiting if anyone complains that the regex does not cover their specific cluster and only then make it configurable (an additional line per scheduler in the config).

**Comment by guillaumeeb** [comment]
> I had a quick look at the infamous ipyparallel file

Yeah its a tough file !

> The controllers all use r'\d+'
> Maybe we can just follow what ipyparallel has done

Doing at least that sounds really good! From what I've seen here, this single regex could cover all our implementation needs. 

FWIW, here are some example of submission outputs:
- PBS
````bash
6757651.admin01
````
- Slurm:
````bash
sbatch: Submitted batch job 65541 

# But with parsable option, it is
65541;cluster
````
- LSF
````bash
Job <930> is submitted to default queue <normal>.
````
- HTCondor (taken somewhere on internet):
````bash
Submitting job(s)con.
Logging submit event(s).
1 job(s) submitted to cluster 6075.

# but there is the -terse option that should make it like
10744.0 - 10744.2
````
- SGE (with correct option, as I understand):
````bash
1234
````

**Comment by guillaumeeb** [comment]
> I am still somewhat unsure about making it configurable

I understand your point, no need to make it configurable if nobody needs it. This is probably true here.

However, in https://github.com/ipython/ipyparallel or https://github.com/jupyterhub/batchspawner, they seem to chose a different approach of making everything configurable if someone ever needs it. Not sure what is the right thing to do. I find it interesting to not being forced to patch or do a PR for every specific case that one can encounter. 

Maybe config will be more interesting and cover most edge cases with #133?

For the time being, having only one `_job_id_from_submit_output` in core.py with a non configurable regex is fine for me.

**Comment by guillaumeeb** [comment]
Closing this as this is resolved in #131
